### ISSUE TITLE
WasapiCapture.Dispose causes freeze

### ISSUE THREAD

**Issue by jfrank14** [naudio/NAudio#630]
My app uses WasapiCapture to record Windows and it nearly always works perfectly. But very, very occasionally, it freezes during the call to Dispose. I checked the source code, and I see this code

```
            if (captureThread != null)
            {
                captureThread.Join();
                captureThread = null;
            }
```

That Join call is a likely candidate for causing a freeze, but I still can't figure out what would cause this, or how to prevent it. It feels like something is causing the thread to remain alive, but because the problem is so rare, it's hard to research further.

Any ideas on what to do about this?

**Comment by markheath** [comment]
are you making sure to always call dispose from the UI main thread?
also, if you can reproduce the issue while debugging, it would be interesting to see what the captureThread is doing at the point

**Comment by jfrank14** [comment]
I'm definitely doing the Dispose on the UI thread. It happens so infrequently that I can't reproduce it on command, or even by trying a bunch of times. I guess my next step would be to fork the WasapiCapture code and add aggressive logging code, looking for some clue, but that feels a bit painful. Any idea why `captureThread` might not be joinable?

**Comment by Erminh** [comment]
Could the `captureThread` be becoming null? Might be worth taking a copy to avoid a `NullReferenceException`

**Comment by jfrank14** [comment]
If it were null, I think it would terminate fine. I can't be sure, but I think something about the thread is stuck, such that the `Join` call never completes.

**Comment by jfrank14** [comment]
After putting this down for a long time, I had to revisit it and I think I found the source of the issue. It turns out that the capture thread is calling a notification callback on the UI thread, which can't respond because the main thread is waiting for the capture thread via the join call. IOW, a deadlock caused by my own code. When I have some time, I will document this in a bit more detail, in case anyone else comes across this issue in the future.
